### PR TITLE
Convert call.js to Typescript & update WebRTC APIs

### PR DESCRIPTION
**Pull request by dbkr**
Two large commits here: a78b590 which converts to typescript and cc4656b which updates to changes in the WebRTC API (mostly using promises instead of callbacks and not using `addStream()`.

This PR also should make the output setting a bit more reliable, as this appears to be quite sensitive on Chrome (it's still not implemented on Firefox).

### CODE REVIEW

File: src/webrtc/call.ts [hunk 1]
```diff
@@ -84,32 +84,73 @@ enum CallParty {
     REMOTE = 'remote',
 }
 
+enum MediaQueueId {
+    REMOTE_VIDEO = 'remote_video',
+    REMOTE_AUDIO = 'remote_audio',
+    LOCAL_VIDEO = 'local_video',
```
Comment (t3chguy) [inline src/webrtc/call.ts:90]: I think our convention is PascalCasing for enum members

File: src/webrtc/call.ts [hunk 2]
```diff
@@ -84,32 +84,73 @@ enum CallParty {
...
+}
+
 enum CallErrorCode {
     /** An error code when the local client failed to create an offer. */
-    ERR_LOCAL_OFFER_FAILED = "local_offer_failed",
+    LOCAL_OFFER_FAILED = 'local_offer_failed',
```
Comment (t3chguy) [inline src/webrtc/call.ts:95]: same re PascalCasing

File: src/webrtc/call.ts [hunk 3]
```diff
@@ -399,8 +367,18 @@ export class MatrixCall extends EventEmitter {
      * @param {Element} element The <code>&lt;video&gt;</code> DOM element.
      */
     setRemoteVideoElement(element : HTMLVideoElement) {
```
Comment (t3chguy) [inline src/webrtc/call.ts:368]: conventionally we don't tend to have a space after the parameter name

File: src/webrtc/call.ts [hunk 4]
```diff
@@ -409,27 +387,41 @@ export class MatrixCall extends EventEmitter {
...
+            return element.play();
+        });
     }
 
     /**
      * Configure this call from an invite event. Used by MatrixClient.
      * @param {MatrixEvent} event The m.call.invite event
      */
-    initWithInvite(event : any) {
+    async initWithInvite(event : any) {
```
Comment (t3chguy) [inline src/webrtc/call.ts:407]: can this get a type?

File: src/webrtc/call.ts [hunk 5]
```diff
@@ -75,8 +75,8 @@ enum CallType {
 }
 
 enum CallDirection {
-    INBOUND,
-    OUTBOUND,
+    INBOUND = 'inbound',
+    OUTBOUND = 'outbound',
```
Comment (t3chguy) [inline src/webrtc/call.ts:79]: Same re PascalCasing

File: src/@types/global.d.ts [hunk 6]
```diff
@@ -36,7 +36,8 @@ declare global {
     }
 
     interface HTMLAudioElement {
-        // setSinkId is experimental and typescript doesn't know about it
+        // sinkId & setSinkId are experimental and typescript doesn't know about it
```
Comment (t3chguy) [inline src/@types/global.d.ts:39]: them?

**Review by t3chguy — COMMENTED** [review]
Looks sane otherwise

File: src/webrtc/call.ts (hunk 2 above)
Comment (dbkr) [reply src/webrtc/call.ts:97]: These ones are part of API unfortunately so need to stay for backwards-compat

File: src/webrtc/call.ts (hunk 4 above)
Comment (dbkr) [reply src/webrtc/call.ts]: do we have a type for events in the js-sdk yet?

File: src/webrtc/call.ts (hunk 2 above)
Comment (dbkr) [reply src/webrtc/call.ts:97]: oh wait, you mean the enum names not the values

File: src/webrtc/call.ts (hunk 4 above)
Comment (t3chguy) [reply src/webrtc/call.ts]: No but you can use `MatrixEvent` (imported) and it'll work magically as soon as that gets Typescripted, helps prevent future issues from cropping up during MatrixEvent being TSified

**Review by t3chguy — APPROVED** [review]

**Pull request merged**